clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -290,18 +317,47 @@ public void poll(long timeout) throws IOException {
 
         clear();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty())
+        boolean dataInBuffers = !keysWithBytesBuffered.isEmpty();
+        
+        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || (madeProgressLastPoll && dataInBuffers))
```

Comment:
there's a (hypothetical) corner case where there's data in the ssl app buffer but the underlying socket is done. this means the socket will never come back from a poll call, and you may wait 300ms for no reason instead of servicing the buffer immediately. this is why dataInBuffers exists.

the made progress flag exists because the downside of the above condition is you may be stuck in a tight loop trying to service the buffer and so we dont try if no progress was made previous attempt.